NimbleMedium/Sources/Domain/Repositories/User/UserSessionRepositoryProtocol.swift

```diff
@@ -0,0 +1,13 @@
+//
+//  UserSessionRepositoryProtocol.swift
+//  NimbleMedium
+//
+//  Created by Minh Pham on 30/08/2021.
+//
+
+import RxSwift
+
+protocol UserSessionRepositoryProtocol: AnyObject {
+
+    func saveUser(_ user: APIUser) -> Completable
```

Comment:
I think in the domain layer, we only need to use the `User` model instead of the `APIUser`
```suggestion
    func saveUser(_ user: User) -> Completable
```

Reply:
@suho `User` is a protocol, it doesn't conform the `Decodable` for storing in the keychain 😅

Reply:
We can convert from `User` to `APIUser` directly in the class `UserSessionRepository`. So the domain layer just needs to know about the abstract model.
```swift
func saveUser(_ user: User) -> Completable {
        Completable.create { [weak self] observer in
            do {
                try self?.keychain.set(KeychainUser(user: user), for: .user)
            } catch {
                observer(.error(error))
            }
            observer(.completed)
            return Disposables.create()
        }
    }
```

What do you think?
Currently, we are following the abstract model approach, so I think we can discuss and improve more about this. And we can write a document about the abstract model. :D

For more detail, as I mentioned in [this document](https://www.notion.so/nimblehq/iOS-Chapter-d1888a31770a44beb0ca52720f4c853b?p=648d73365c3f42eeb89e618f609f2261), the domain layer has no dependency on other layers. So it should use `User` model instead of `APIUser`. The domain layer should not care about `API` or `Keychain`, so we can update the implementation of UserSessionRepositoryProtocol more easily.

What do you think?

Reply:
I think it is a nice suggestion, let me try and apply 😁